pdf/pdfselection.cpp

```diff
@@ -0,0 +1,447 @@
+        d->pageIndexStop = pageIndex;
+        d->boxIndexStop = boxIndex;
+        endInsertRows();
+    } else {
+        beginRemoveRows(QModelIndex(), count_ + nBoxes, count_ - 1);
+        d->pageIndexStop = pageIndex;
+        d->boxIndexStop = boxIndex;
+        endRemoveRows();
+    }
+    emit countChanged();
+    emit textChanged();
+
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+}
+
+void PDFSelection::selectAt(const QPointF &point)
+{
+    unselect();
+
+    int pageIndex, boxIndex;
+    d->textBoxId(point, PDFSelection::Private::At, &pageIndex, &boxIndex);
+    if (pageIndex < 0 || boxIndex < 0)
+        return;
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(pageIndex);
+    QRectF box = boxes[boxIndex].first;
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    d->pageIndexStart = pageIndex;
+    d->boxIndexStart = boxIndex;
+    d->pageIndexStop = pageIndex;
+    d->boxIndexStop = boxIndex;
+    endInsertRows();
+    emit countChanged();
+    emit textChanged();
+
+    d->start = QPointF(box.x(), box.y() + box.height() / 2.);
+    emit startChanged();
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+}
+
+void PDFSelection::unselect()
+{
+    beginResetModel();
+    d->pageIndexStart = d->pageIndexStop = -1;
+    d->boxIndexStart = d->boxIndexStop = -1;
+    endResetModel();
+    emit countChanged();
+    emit textChanged();
+}
+
+void PDFSelection::onLayoutChanged()
+{
+    emit dataChanged(createIndex(0, 0), createIndex(count() - 1, 0), QVector<int>{Rect});
+}
+
+bool PDFSelection::selectionAtPoint(const QPointF &point) const
+{
+    if (d->pageIndexStart < 0 || d->pageIndexStop < 0 ||
+        d->boxIndexStart < 0 || d->boxIndexStop < 0)
+        return false;
+
+    if (!d->canvas)
+        return false;
+    PDFDocument *doc = d->canvas->document();
+    if (!doc)
+        return false;
+    
+    int i;
+    QPair<int, QRectF> at = d->canvas->pageAtPoint(point);
+    if (at.first < 0)
+        return false;
+    QPointF reducedCoordPoint {point.x() / at.second.width(),
+            (point.y() - at.second.y()) / at.second.height()};
+    const PDFDocument::TextList &boxes = doc->textBoxesAtPage(at.first);
+    for (i = ((at.first == d->pageIndexStart) ? d->boxIndexStart : 0);
+         i < ((at.first == d->pageIndexStop) ? d->boxIndexStop + 1 : boxes.length());
+         i++) {
+        if (boxes.value(i).first.contains(reducedCoordPoint))
+            return true;
+    }
+    return false;
+}
+
+QString PDFSelection::text() const
+{
+    QString out;
+
+    if (d->pageIndexStart < 0 || d->pageIndexStop < 0 ||
+        d->boxIndexStart < 0 || d->boxIndexStop < 0)
+        return out;
+
+    if (!d->canvas)
+        return out;
+    PDFDocument *doc = d->canvas->document();
+    if (!doc)
+        return out;
+    
+    out = QLatin1String("");
```

Comment:
Oh right, it's already declared QString. Then again this would be then more or less useless line?